sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -0,0 +1,413 @@
+        if (gridfs != null) {
+          gridfs = null;
+          mongo.close();
+          mongo = null;
+        }
+      }
+
+      @Override
+      public List<? extends BoundedSource<T>> splitIntoBundles(long desiredBundleSizeBytes,
+          PipelineOptions options) throws Exception {
+        try {
+          setupGridFS();
+          DBCursor cursor;
+          if (filterJson != null) {
+            DBObject query = (DBObject) JSON.parse(filterJson);
+            cursor = gridfs.getFileList(query).sort(null);
+          } else {
+            cursor = gridfs.getFileList().sort(null);
+          }
+                    long size = 0;
+          List<BoundedGridFSSource<T>> list = new LinkedList<>();
+          List<ObjectId> objects = new LinkedList<>();
+          while (cursor.hasNext()) {
+            GridFSDBFile file = (GridFSDBFile) cursor.next();
+            long len = file.getLength();
+            if ((size + len) > desiredBundleSizeBytes && !objects.isEmpty()) {
+              list.add(new BoundedGridFSSource<T>(uri, database, bucket, filterJson,
+                                                  parser, coder, objects));
+              size = 0;
+              objects = new LinkedList<>();
+            }
+            objects.add((ObjectId) file.getId());
+            size += len;
+          }
+          if (!objects.isEmpty() || list.isEmpty()) {
+            list.add(new BoundedGridFSSource<T>(uri, database, bucket, filterJson,
+                                                parser, coder, objects));
+          }
+          return list;
+        } finally {
+          closeGridFS();
+        }
+      }
+      @Override
+      public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+        try {
+          setupGridFS();
+          DBCursor cursor;
+          if (filterJson != null) {
+            DBObject query = (DBObject) JSON.parse(filterJson);
+            cursor = gridfs.getFileList(query).sort(null);
+          } else {
+            cursor = gridfs.getFileList().sort(null);
+          }
+          long size = 0;
+          while (cursor.hasNext()) {
+            GridFSDBFile file = (GridFSDBFile) cursor.next();
+            size += file.getLength();
+          }
+          return size;
+        } finally {
+          closeGridFS();
+        }
+      }
+
+      @Override
+      public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+        return false;
+      }
+      @Override
+      public org.apache.beam.sdk.io.BoundedSource.BoundedReader<T> createReader(
+          PipelineOptions options) throws IOException {
+        return new GridFSReader(this);
+      }
+
+
+      @Override
+      public void validate() {
+      }
+
+      @SuppressWarnings("unchecked")
+      @Override
+      public Coder<T> getDefaultOutputCoder() {
+        if (coder != null) {
+          return coder;
+        }
+        return (Coder<T>) SerializableCoder.of(Serializable.class);
+      }
+
```

Comment:
The source should also contain `populateDisplayData()` method. This method is used by the runner to "display" the source configuration (for instance in an UI).
You should add something like:
```
    public void populateDisplayData(DisplayData.Builder builder) {
      builder.add(DisplayData.item("bucket", bucket));
      builder.add(DisplayData.item("uri", uri));
      builder.addIfNotNull(DisplayData.item("jsonFilter", jsonFilter));
      ....
   }
```